### ISSUE TITLE
add edge norm to EGAT and GAT 

### ISSUE THREAD

**Issue by venzino-han** [dmlc/dgl#5459]
## 🚀 Feature
Add edge norm to EGAT and GAT for edge masking. 

## Motivation
EGAT and GAT calculate attention weight between nodes. 
However, in some cases we need to mask some edges and need to block message passing.
There may be cases where we need to filter only certain kinds of edges for message passing, 
or where we want to artificially increase or decrease the attribute weight based on edge type or node type.

For example, I used two EGAT layers. I created two layers, one utilizing local edge attachments and one utilizing global edge attachments. When utilizing local edge attachments, I masked the global edges, and when utilizing global edge attachments, I did the opposite.

## Alternatives
I applied the edge norm to the forward function to reflect the edge norm in the EGAT and GAT source code.

## Pitch
```
def forward(self, graph, feat, norm=None, get_attention=False)
```
Add a `norm` parameter to the forward function of each model.
The `norm` parameter is multiplied by the attention weight before message passing is performed. 

## Additional context
I needed this feature for my thesis and implemented it myself. I hope to be able to contribute myself.

**Comment by venzino-han** [comment]
Hi @jermainewang, thanks for your kind reply.
Here is my answer to your question.
- What is the expected type and shape of the norm argument?
The expected type of `norm` is torch.Tensor (1D tensor) and the shape would be |𝐸|,).
(same as `norm` of `RelGraphConv`)
- Will the norm array be multiplied to the unnormalized or normalized attention weights?
I applied norm after the softmax. However, I just used norm to mask the edges.

**Comment by mufeili** [comment]
Is this equivalent to passing graphs with certain edges removed from the original graph?

**Comment by venzino-han** [comment]
@mufeili That can be one of the use case. 
However, in my case, I used multiple EGAT layers and the masked edges should be used in the next layers. 

Another use case is when you want to artificially adjust the attention coefficient for each node type, in which case we can multiply each attention coefficient with norm tensor.

**Comment by mufeili** [comment]
> However, in my case, I used multiple EGAT layers and the masked edges should be used in the next layers.

I don't understand why this is an issue. If you want to mask the same set of edges in all these EGAT layers, you can first have a variant of the original graph with the particular edges removed and then use it in all these EGAT layers.

If you really want to rescale the computed attention coefficients, you may consider following the practice of the `edge_weight` argument as in [this module](https://docs.dgl.ai/generated/dgl.nn.pytorch.conv.GraphConv.html) and make a contribution for `GATConv` and `EGATConv`.

**Comment by venzino-han** [comment]
@mufeili  I understand what you mean. Thanks for the suggestion. I would like to contribute as your suggestion.
I will let you know when I create PR.

**Comment by mufeili** [comment]
@venzino-han Sounds great!
